syntax = "proto3";

package saasforge.payment;

option go_package = "github.com/saasforge/proto/payment";

service PaymentService {
  rpc CreateSubscription(CreateSubscriptionRequest) returns (SubscriptionResponse);
  rpc UpdateSubscription(UpdateSubscriptionRequest) returns (SubscriptionResponse);
  rpc CancelSubscription(CancelSubscriptionRequest) returns (SubscriptionResponse);
  rpc GetSubscription(GetSubscriptionRequest) returns (SubscriptionResponse);
  rpc AddPaymentMethod(AddPaymentMethodRequest) returns (PaymentMethodResponse);
  rpc RemovePaymentMethod(RemovePaymentMethodRequest) returns (RemovePaymentMethodResponse);
  rpc GetInvoice(GetInvoiceRequest) returns (InvoiceResponse);
  rpc RecordUsage(RecordUsageRequest) returns (RecordUsageResponse);
}

enum SubscriptionStatus {
  SUBSCRIPTION_STATUS_UNSPECIFIED = 0;
  ACTIVE = 1;
  PAST_DUE = 2;
  CANCELED = 3;
  UNPAID = 4;
  TRIALING = 5;
}

message CreateSubscriptionRequest {
  string tenant_id = 1;
  string plan_id = 2;
  string payment_method_id = 3;
  int32 quantity = 4;
  optional int32 trial_days = 5;
  string idempotency_key = 6;
}

message UpdateSubscriptionRequest {
  string tenant_id = 1;
  string subscription_id = 2;
  optional string plan_id = 3;
  optional int32 quantity = 4;
}

message CancelSubscriptionRequest {
  string tenant_id = 1;
  string subscription_id = 2;
  bool immediate = 3;
}

message GetSubscriptionRequest {
  string tenant_id = 1;
  string subscription_id = 2;
}

message SubscriptionResponse {
  string id = 1;
  string tenant_id = 2;
  string plan_id = 3;
  SubscriptionStatus status = 4;
  int64 current_period_start = 5;
  int64 current_period_end = 6;
  optional int64 cancel_at = 7;
  int32 quantity = 8;
  double mrr = 9;
}

message AddPaymentMethodRequest {
  string tenant_id = 1;
  string stripe_payment_method_id = 2;
}

message PaymentMethodResponse {
  string id = 1;
  string type = 2;
  string last4 = 3;
  optional string brand = 4;
  optional int32 exp_month = 5;
  optional int32 exp_year = 6;
}

message RemovePaymentMethodRequest {
  string tenant_id = 1;
  string payment_method_id = 2;
}

message RemovePaymentMethodResponse {
  bool success = 1;
}

message GetInvoiceRequest {
  string tenant_id = 1;
  string invoice_id = 2;
}

message InvoiceResponse {
  string id = 1;
  string tenant_id = 2;
  string subscription_id = 3;
  double amount_due = 4;
  double amount_paid = 5;
  string status = 6;
  int64 due_date = 7;
  optional int64 paid_at = 8;
  optional string pdf_url = 9;
}

message RecordUsageRequest {
  string tenant_id = 1;
  string subscription_id = 2;
  string metric_name = 3;
  int64 quantity = 4;
  int64 timestamp = 5;
}

message RecordUsageResponse {
  bool success = 1;
  string usage_record_id = 2;
}
